language-plutus-core/src/Language/PlutusCore/Name.hs

```diff
@@ -85,39 +84,40 @@ newtype Unique = Unique { unUnique :: Int }
 -- | The unique of a type-level name.
 newtype TypeUnique = TypeUnique
     { unTypeUnique :: Unique
-    }
+    } deriving (Eq)
 
 -- | The unique of a term-level name.
 newtype TermUnique = TermUnique
     { unTermUnique :: Unique
-    }
-
--- | The default implementation of 'HasUnique' for newtypes.
-newtypeUnique
-    :: (Wrapped new, HasUnique (Unwrapped new) unique', Coercible unique' unique)
-    => Lens' new unique
-newtypeUnique = _Wrapped' . unique . coerced
+    } deriving (Eq)
 
 -- | Types which have a 'Unique' attached to them, mostly names.
-class Coercible Unique unique => HasUnique a unique | a -> unique where
+class Coercible unique Unique => HasUnique a unique | a -> unique where
     unique :: Lens' a unique
+    -- | The default implementation of 'HasUnique' for newtypes.
+    default unique
+        :: (Wrapped a, HasUnique (Unwrapped a) unique', Coercible unique' unique)
+        => Lens' a unique
+    unique = _Wrapped' . unique . coerced
+
+instance HasUnique Unique Unique where
+    unique = id
 
 instance HasUnique (Name ann) TermUnique where
     unique = lens g s where
         g = TermUnique . nameUnique
         s n (TermUnique u) = n{nameUnique=u}
 
-instance HasUnique (TyName ann) TypeUnique where
-    unique = newtypeUnique
+instance HasUnique (TyName ann) TypeUnique
 
 -- | A mapping from uniques to values of type @a@.
 newtype UniqueMap unique a = UniqueMap
     { unUniqueMap :: IM.IntMap a
     } deriving newtype (Show, Semigroup, Monoid, Functor)
 
 -- | Insert a value by a unique.
-insertByUnique :: Coercible Unique unique => unique -> a -> UniqueMap unique a -> UniqueMap unique a
-insertByUnique uniq x = UniqueMap . IM.insert (coerce uniq) x . unUniqueMap
+insertByUnique :: Coercible unique Unique => unique -> a -> UniqueMap unique a -> UniqueMap unique a
+insertByUnique uniq = coerce . IM.insert (coerce uniq)
```

Comment:
Hm, inspired by an earlier example of yours, could this entirely be `insertByUnique = coerce IM.insert`?

Reply:
Nope, doesn't type check. `coerce` works until it doesn't.